ext/gctrack/gctrack.c

```diff
@@ -4,184 +4,162 @@
-        cur = cur->next;
+  if (!enabled) {
+    return;
+  }
+  rb_trace_arg_t *tparg = rb_tracearg_from_tracepoint(tpval);
+  switch (rb_tracearg_event_flag(tparg)) {
+    case RUBY_INTERNAL_EVENT_GC_ENTER: {
+      last_enter = nanotime();
+    }
+      break;
+    case RUBY_INTERNAL_EVENT_GC_EXIT: {
+      if (last_enter) {
+        add_gc_cycle(nanotime() - last_enter);
       }
+      last_enter = 0;
     }
+      break;
   }
-  return ev;
 }
 
-static void
-flush_gc_events(void *arg)
+static VALUE
+gctracker_start_record(int argc, VALUE *argv, VALUE klass)
 {
-  gctrack_hook_data_t *hook_data = arg;
+  if(!enabled) {
+    return Qfalse;
+  }
 
-  // TODO: convert gctrack_events to ruby array for the callback
-  rb_proc_call(hook_data->proc, rb_ary_new());
+  record_t *record = (record_t *) calloc(1, sizeof(record_t));
+  if (!record) {
+    return Qfalse;
+  }
+  
+  record->parent = last_record;
+  last_record = record;
+  return Qtrue;
 }
 
-static void
-gc_event_hook(rb_event_flag_t event, VALUE data, VALUE self, ID id, VALUE klass)
+static VALUE
+gctracker_end_record(int argc, VALUE *argv, VALUE klass)
 {
-  gctrack_hook_data_t *hook_data;
-  gctrack_event_t *ev;
-  bool first = false;
-
-  TypedData_Get_Struct(data, gctrack_hook_data_t, &gctrack_hook_data_type, hook_data);
-  ev = alloc_event(hook_data, &first);
-
-  clock_gettime(CLOCK_MONOTONIC_RAW, &ev->ts);
-  ev->ev = event;
-
-  if (first) {
-    rb_postponed_job_register(0, flush_gc_events, hook_data);
-  }
+  if (!last_record) {
+    return Qnil;
+  } 
+  record_t *record = last_record;
+  last_record = record->parent;
+  
+  VALUE stats = rb_ary_new2(2);
+  rb_ary_store(stats, 0, ULONG2NUM(record->cycles));
+  rb_ary_store(stats, 1, ULONG2NUM(record->duration));
+
+  free(record);  
+
+  return stats;
 }
 
 static VALUE
-enable(int argc, VALUE *argv, VALUE klass)
+gctracker_enable(int argc, VALUE *argv, VALUE klass)
 {
-  gctrack_hook_data_t *data;
-  VALUE obj;
-  VALUE opts;
-  VALUE ctx = Qnil;
-
-  rb_event_flag_t events = RUBY_INTERNAL_EVENT_GC_ENTER |
-    RUBY_INTERNAL_EVENT_GC_EXIT;
-
-  if (enabled) {
-    return Qfalse;
-  }
-
-  if (!rb_block_given_p()) {
-    rb_raise(rb_eArgError, "called without a block");
-  }
-
-  rb_scan_args(argc, argv, ":", &opts);
-  if (!NIL_P(opts)) {
-    ID keys[2];
-    VALUE values[2];
-
-    keys[0] = rb_intern("context");
-    keys[1] = rb_intern("events");
-    rb_get_kwargs(opts, keys, 0, 2, values);
-    if (values[0] != Qundef) {
-      ctx = values[0];
+  VALUE tracepoint = rb_ivar_get(cTracker, id_tracepoint);
+  if (NIL_P(tracepoint)) {
+    enabled = true;
+    rb_event_flag_t events;
+    events = RUBY_INTERNAL_EVENT_GC_ENTER | RUBY_INTERNAL_EVENT_GC_EXIT;
+    tracepoint = rb_tracepoint_new(0, events, gctracker_hook, (void *) NULL);
+    if (NIL_P(tracepoint)) { 
+      enabled = false;
+      return Qfalse;
     }
+    rb_ivar_set(cTracker, id_tracepoint, tracepoint);
+    rb_tracepoint_enable(tracepoint);
```

Comment:
You should probably check the return value here.

Reply:
`rb_tracepoint_enable` ? That returns `Qundef` always, right?